### PR TITLE
Assert that DOM structs have the correct first field

### PR DESCRIPTION
**Pull request by Manishearth**
Not having the right field can lead to fun bugs like https://github.com/ferjm/servo/pull/1 where the struct gets mis-reinterpreted as what should be its parent (but is not layout-wise)

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21105)

### CODE REVIEW

**Comment by KiChjang** [comment]
Huh, I thought that we already perform this check... Perhaps it isn't robust enough and it only checks whether the first field is also a DOM struct?

**Comment by nox** [comment]
Please describe what the changes do in the commit message, and maybe show what is generated here. I don't like how there is a special case for some canvas-related interface in the middle of codegen with this PR.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 1]
```diff
@@ -2161,8 +2165,47 @@ def define(self):
...
+            # also has a reflector
+            parentName = "::dom::canvasrenderingcontext2d::CanvasRenderingContext2D"
+        args = {
+            "parentName": parentName,
+            "selfName": selfName,
+        }
+
+        return """\
+impl %(selfName)s {
+    fn __assert_parent_type(&self) {
```
Comment (nox) [inline components/script/dom/bindings/codegen/CodegenRust.py]: I would rather encode the relation with a trait specifying from which interface it is extended.

**Comment by nox** [comment]
Also please don't write `r?` in PR descriptions.

**Comment by Manishearth** [comment]
Regarding canvas, I discussed this with @asajeffrey , who had written that, and for efficiency the paint rendering context struct internally "inherits" from the canvas rendering context, despite not webidl inheriting. This is still safe because canvas eventually inherits from EventTarget, so the layouts work out.

Regarding `r?` , I thought that was only for when the sync service was running, which it isn't anymore.

File: components/script/dom/bindings/codegen/CodegenRust.py (hunk 1 above)
Comment (Manishearth) [reply components/script/dom/bindings/codegen/CodegenRust.py]: You still need to assert the existence of this impl somehow, though. Thoughts?

**Comment by nox** [comment]
> Regarding canvas, I discussed this with @asajeffrey, who had written that, and for efficiency the paint rendering context struct internally "inherits" from the canvas rendering context, despite not webidl inheriting. This is still safe because canvas eventually inherits from EventTarget, so the layouts work out.

You can use `[Inline]` on a fake WebIDL interface, like I did for `Window` and `WorkerGlobalScope` which both inherit from `GlobalScope` which doesn't actually exist on the JS side.

> Regarding `r?`, I thought that was only for when the sync service was running, which it isn't anymore.

Oh right.

**Comment by Manishearth** [comment]
Created a HasParent trait and fixed the commit message. Left a FIXME for the `[Inline]`

@bors r=nox

**Pull request merged**